tenets/codelingo/php/sql-concats/ideal-solution.lingo

```diff
@@ -1,62 +1,127 @@
+                } else {
+                  $safeByte += 1
+                }
+              }
     flows:
       codelingo/review:
         comment: SQL query strings shouldnt be concatenated as these are susceptible to SQL injection
     query: |
       import codelingo/ast/php
 
-      # Find values from an http request
+      # Find a value extracted from an http request.
       php.stmt_classmethod(depth = any):
         php.expr_assign:
-          php.expr_variable:
-            name as httpVariable
           php.expr_methodcall(depth = any): # PHP surprisingly appears to have objects hanging off method calls
-            name == "get"
+            name == "get" #TODO: handle other http verbs - ideally more robustly
             php.expr_propertyfetch:
               name == "request"
 
-        # Find method that concatenates the http value into a string
-        php.calls(depth = any):
-          php.arg:
-            name as httpVariable
-          # Recurse along method call chain to find concatenation method
-          # TODO: validate that the $httpVariable is passed along call chain
-          php.calls(depth = any, follow: callgraph):
-            php.stmt_classmethod(depth = any):
-              php.param:
-                # find variables whose value can be affected by the parameter
-                php.may_assign_to:
-                  name as httpArg
-              any_of:
-                php.expr_assignop_concat(depth = any):
-                  @ review.comment
-                  php.expr_variable(depth = any):
-                    name as sqlStatement
-                    php.expr_variable(depth = any):
-                      name as httpArg
-                php.expr_assign(depth = any):
-                  @ review.comment
-                  php.expr_variable(depth = any):
-                    name as sqlStatement
-                    php.expr_variable(depth = any):
-                      name as httpArg
-                  php.expr_binaryop_concat(depth = any)
-
-              # Check that $sqlStatement is passed to a database method
-              @ review.comment
-              php.calls(depth = any):
-                php.arg:
+          # Follow the reaches graph to function calls.
+          # Stop at variables defined by a call to the sanatise function.
+          path(repeat = 0:):
+            edge("reaches"):
+              php.expr_assign:
+                exclude:
+                  php.expr_methodcall(depth = any):
+                    name == "sanatise"
+                    php.expr_propertyfetch:
+                      name == "db"
+                pathcontinue
+          pathend:
+            edge("defines"):
+              php.expr_variable:
+                id as callVar1
+        php.method_call(depth = any): # fork path to reach
+          php.expr_variable:
+            id == callVar1
+          # Recursively traverse to the function definition.
+          path(repeat = 0:):
```

Comment:
What would it mean to repeat a path starting at >0?

Reply:
It may be better to use a single integer for repeat and `-1` to indicate `forever`.

Reply:
Repeat is how many times you repeat the path. `(repeat = 0:)` means "repeat zero or more times". We could potentially expand it to `(repeat = 0:-1)` or `(repeat = 0:any)` if you think that's clearer - it's based on slice syntax right now, i.e., https://play.golang.org/p/4OXuSZNsc8J.

`(repeat = 1:)` is another important use case which means "repeat one or more times". I think it's important to have `(repeat = n:)` for n > 1 as well, because you could have a problem with 4 levels of nested if statements, or something. So it think it's important to be able to specify a range.